src/Validation.cpp

```diff
@@ -290,291 +454,147 @@ void ValidateVariableTypeVisitor::visitNamedType(const peg::ast_node& namedType)
 		return;
 	}
 
-	response::Value kind(response::Type::EnumValue);
-
-	switch (itrKind->second)
-	{
-		case introspection::TypeKind::SCALAR:
-			kind.set<response::StringType>(R"gql(SCALAR)gql");
-			break;
-
-		case introspection::TypeKind::ENUM:
-			kind.set<response::StringType>(R"gql(ENUM)gql");
-			break;
-
-		case introspection::TypeKind::INPUT_OBJECT:
-			kind.set<response::StringType>(R"gql(INPUT_OBJECT)gql");
-			break;
-
-		default:
-			return;
-	}
-
-	_isInputType = true;
-	_variableType.emplace_back(R"gql(kind)gql", std::move(kind));
-	_variableType.emplace_back(R"gql(name)gql", response::Value(std::move(name)));
+	_variableType = std::make_shared<ValidateType>(itrKind->second, std::move(name));
 }
 
 void ValidateVariableTypeVisitor::visitListType(const peg::ast_node& listType)
 {
-	response::Value kind(response::Type::EnumValue);
 	ValidateVariableTypeVisitor visitor(_typeKinds);
 
-	kind.set<response::StringType>(R"gql(LIST)gql");
 	visitor.visit(*listType.children.front());
-	_isInputType = visitor.isInputType();
-	_variableType.emplace_back(R"gql(kind)gql", std::move(kind));
-	_variableType.emplace_back(R"gql(ofType)gql", visitor.getType());
+
+	_variableType =
+		std::make_shared<ValidateType>(introspection::TypeKind::LIST, visitor.getType());
 }
 
 void ValidateVariableTypeVisitor::visitNonNullType(const peg::ast_node& nonNullType)
 {
-	response::Value kind(response::Type::EnumValue);
 	ValidateVariableTypeVisitor visitor(_typeKinds);
 
-	kind.set<response::StringType>(R"gql(NON_NULL)gql");
 	visitor.visit(*nonNullType.children.front());
-	_isInputType = visitor.isInputType();
-	_variableType.emplace_back(R"gql(kind)gql", std::move(kind));
-	_variableType.emplace_back(R"gql(ofType)gql", visitor.getType());
+	_variableType =
+		std::make_shared<ValidateType>(introspection::TypeKind::NON_NULL, visitor.getType());
 }
 
 bool ValidateVariableTypeVisitor::isInputType() const
 {
-	return _isInputType;
+	return _variableType && _variableType->isInputType();
 }
 
-ValidateType ValidateVariableTypeVisitor::getType()
+std::shared_ptr<ValidateType>&& ValidateVariableTypeVisitor::getType()
 {
-	auto result = std::move(_variableType);
+	return std::move(_variableType);
+}
 
-	return result;
+ValidationContext::ValidationContext(const Request& service)
+{
+	// TODO: we should execute this query only once per schema,
+	// maybe it can be done and cached inside the Request itself to allow
+	// this. Alternatively it could be provided at compile-time such as schema.json
+	// that is parsed, this would allow us to drop introspection from the Request
+	// and still have it to work
+	auto ast = peg::parseString(introspectionQuery);
+	// This is taking advantage of the fact that during validation we can choose to execute
+	// unvalidated queries against the Introspection schema. This way we can use fragment
+	// cycles to expand an arbitrary number of wrapper types.
+	ast.validated = true;
```

Comment:
Using the canonical `IntrospectionQuery` does have this limitation, where it can only follow links to a fixed depth. It's compatible with more tools, but it does put a limit on the complexity of the schema.

Reply:
I get this, but while there may be a combination of lists, non-null and the actual type, it's pretty unusual in real life, even more that it would not work in any other GraphQL tool.

Do you want me to move and use the non-standard recursive here?